### PR TITLE
refactor ca.nrc.cadc.net Http classes

### PR DESCRIPTION
**Pull request by pdowler**
- extend error handling to map http response code to exceptions (reverse of cadc-rest)
- add HttpTransfer.prepare() method that performs action (prepares the input stream) and throws
- run() still prepares, reads response, and captures exceptions

### CODE REVIEW

File: cadc-util/src/main/java/ca/nrc/cadc/net/HttpDownload.java [hunk 1]
```diff
@@ -458,27 +388,26 @@ private void runX()
             log.debug("caught: " + tex);
             throwTE = true;
             throw tex;
-        } catch (AccessControlException ex) {
+        } catch (AccessControlException | ExpectationFailedException | IllegalArgumentException | PreconditionFailedException | ResourceNotFoundException ex) {
```
Comment (brianmajor) [inline cadc-util/src/main/java/ca/nrc/cadc/net/HttpDownload.java]: I'm sure there's a good reason for this set of 400 level errors to be caught here but I don't know what that is.

File: cadc-util/src/main/java/ca/nrc/cadc/net/ExpectationFailedException.java [hunk 2]
```diff
@@ -65,117 +65,24 @@
...
-import ca.nrc.cadc.db.DBUtil;
-import ca.nrc.cadc.util.Log4jInit;
-import javax.sql.DataSource;
-import org.apache.log4j.Level;
 import org.apache.log4j.Logger;
-import org.junit.Assert;
-import org.junit.Test;
 
 /**
  *
```
Comment (brianmajor) [inline cadc-util/src/main/java/ca/nrc/cadc/net/ExpectationFailedException.java]: I find it nice to have a brief description of what the exception means in the class level javadoc.

File: cadc-util/src/main/java/ca/nrc/cadc/net/HttpDownload.java (hunk 1 above)
Comment (pdowler) [reply cadc-util/src/main/java/ca/nrc/cadc/net/HttpDownload.java]: removed these catch blocks

File: cadc-util/src/main/java/ca/nrc/cadc/net/ExpectationFailedException.java (hunk 2 above)
Comment (pdowler) [reply cadc-util/src/main/java/ca/nrc/cadc/net/ExpectationFailedException.java]: added one liner (cut and paste from http definition of 417)

**Comment by pdowler** [comment]
Before merging, I am going to put some thought into which exceptions should be declared RuntimeException and which are (checked) Exception... I think anything that a client could know a priori will never happen because it doesn't do the thing that might trigger it should be unchecked (Runtime) so the client does not have to declare it in the throws. That was a client throws clause will actually be meaningful.

**Pull request merged**